File: 0001-0500/0025. Reverse Nodes in k-Group.java
// Hard
// LinkedList
// Time:O(n), Space:O(n/k)
// https://leetcode.cn/problems/reverse-nodes-in-k-group/

class Solution{
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) return null;

        // 找区间
        ListNode start = head, end = head;
        for (int i = 0; i < k; i++) {
            if (end == null) return head;
            end = end.next;
        }

        // 反转区间: 返回反转后的头节点
        ListNode newHead = reverse(start, end);
        
        // 和原本的接上：原本的结尾和前一段的最后一个接上
        start.next = reverseKGroup(end, k);
        return newHead;
    }
    
    private ListNode reverse(ListNode start, ListNode end) {
        ListNode pre, cur, next;
        pre = null;
        cur = start;
        next = start;
        while (cur != end) {
            next = cur.next;
            cur.next = pre; // 反转当前节点的指针
            pre = cur;
            cur = next;
        }
        return pre;
    }
}

File: 0001-0500/0025. Reverse Nodes in k-Group.py
from typing import *

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def hasKNodes(node: Optional[ListNode], k: int) -> bool:
            count = 0
            while node and count < k:
                node = node.next
                count += 1
            return count == k
        
        def reverse(head: Optional[ListNode], k: int) -> Optional[ListNode]:
            pre = None
            cur = head
            for _ in range(k):
                next_node = cur.next
                cur.next = pre
                pre = cur
                cur = next_node
            return pre
        
        dummy = ListNode(0)
        dummy.next = head
        pre_group = dummy

        while hasKNodes(pre_group.next, k):
            group_start = pre_group.next
            group_end = group_start
            for _ in range(k - 1):
                group_end = group_end.next
            
            next_group = group_end.next
            new_head = reverse(group_start, k)

            pre_group.next = new_head
            group_start.next = next_group
            pre_group = group_start
        
        return dummy.next
            